### PR TITLE
Adding throwing and AdaptivePerformance tools 

### PR DESCRIPTION
**Pull request by keluecke**
Overview
---
See [optimizing performance for unity apps](https://developer.microsoft.com/en-us/windows/mixed-reality/performance_recommendations_for_immersive_headset_apps#optimizing_performance_for_unity_apps) (should be updated soon) and [throwing objects](https://developer.microsoft.com/en-us/windows/mixed-reality/gestures_and_motion_controllers_in_unity#throwing_objects) for details.

### CODE REVIEW

File: Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    public GameObject[] ThrowingAssets;
+    public InteractionSourceNode ControllerPose = InteractionSourceNode.Grip;
+    public Transform RealWorldRoot;
+
+    private readonly Dictionary<uint, Transform> devices = new Dictionary<uint, Transform>();
+    private readonly Dictionary<uint, int> modelIndecies = new Dictionary<uint, int>();
+    private readonly Dictionary<uint, bool> isDetatched = new Dictionary<uint, bool>();
+
+    // Use this for initialization
+    void Start ()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs]: Missing explicit private access modifiers for unity api's and formatting is off.
We can probably lose the comment too.

File: Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    {
+		if (ThrowingAssets.Length == 0)
+        {
+            throw new System.Exception("ThrowingSceneManager needs to have some throwable objects!");
+        }   
+
+        InteractionManager.InteractionSourceReleased += InteractionManager_InteractionSourceReleased;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs]: We should probably be using the input manager to get our interactions.

File: Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs [hunk 3]
```diff
@@ -18,6 +18,10 @@ public class PlaneTargetGroupPicker : Singleton<PlaneTargetGroupPicker>
 
         private Coroutine displayForSecondsCoroutine;
 
+        protected override void InitializeInternal()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs]: What's with this added override method that got added to all these irrelevant classes to this PR?

File: Assets/HoloToolkit/Utilities/Scripts/GpuTiming/GpuTimingCamera.cs [hunk 4]
```diff
@@ -24,15 +28,15 @@ protected void OnPreRender()
 		{
 			if (timingCamera.stereoActiveEye == Camera.MonoOrStereoscopicEye.Left || timingCamera.stereoActiveEye == Camera.MonoOrStereoscopicEye.Mono)
 			{
-				GpuTiming.BeginSample(TimingTag);
+                GpuTiming.BeginSample(TimingTag);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/GpuTiming/GpuTimingCamera.cs]: strange formatting changes here.

File: Assets/HoloToolkit/Utilities/Scripts/Performance/ShaderControl.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+        private int currentShaderLevel = 0;
+        private Renderer objectRenderer;
+
+        private void Awake()
+        {
+            objectRenderer = GetComponent<Renderer>();
+            if (materialList.Length > 0 && materialList[0] == null)
+            {
+                materialList[0] = objectRenderer.material;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/ShaderControl.cs]: This is a very bad documented memory leak in Unity.  We should also be destroying the `materialList` material instances we make here in OnDestory.

File: Assets/HoloToolkit/Utilities/Scripts/Performance/ViewportScaleManager.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            {
+                voiceControlledViewport = 1.0f;
+            });
+
+            keywords.Add("Set viewport to point zero five", () =>
+            {
+                voiceControlledViewport = 1.0f;
+            });
+
+            keywordRecognizer = new KeywordRecognizer(keywords.Keys.ToArray());
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/ViewportScaleManager.cs]: We should probably use the ISpeechHandler instead of creating another instance of the keyword recognizer.

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 7]
```diff
@@ -26,20 +27,29 @@ public static T Instance
...
+                    return instance;
+                }
+
+                if (!searchForInstance)
+                {
+                    return null;
+                }
+
+                // likly first time
+                T[] objects = FindObjectsOfType<T>();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: If we use [`Resources.FindObjectsOfTypeAll<T>();`](https://docs.unity3d.com/ScriptReference/Resources.FindObjectsOfTypeAll.html) we can find disabled GameObject instances as well.

> Contrary to Object.FindObjectsOfType this function will also list disabled objects

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 8]
```diff
@@ -11,7 +11,8 @@ namespace HoloToolkit.Unity
     /// parent root GameObject with <see cref="Object.DontDestroyOnLoad"/></remarks>
     /// </summary>
     /// <typeparam name="T">The Singleton Type</typeparam>
-    public class Singleton<T> : MonoBehaviour where T : Singleton<T>
+    [DisallowMultipleComponent]
+    public abstract class Singleton<T> : MonoBehaviour where T : Singleton<T>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: I think this change is outside the scope of this PR and should be addressed separately.

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs.meta [hunk 9]
```diff
@@ -1,8 +1,9 @@
 fileFormatVersion: 2
-guid: b0ed29a95096a25448535f8df9cf0423
-timeCreated: 1455735878
+guid: 3ac3ff05b2d583947859298e1ee17955
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs.meta]: How does changing this guid affect the references in other people's projects?  Probably best to revert.

File: Assets/HoloToolkit/Utilities/Scripts/Throwing.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.XR;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Throwing.cs]: Not properly wrapped to support VR namespace

File: Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs (hunk 2 above)
Comment (keluecke) [reply Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs]: I want to minimize dependencies, as this is just a sample, allowing users to add throwing without needing to dig through the whole input stack

File: Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs (hunk 3 above)
Comment (keluecke) [reply Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs]: This is introducing an abstract initialization step in singleton to help auto sort out dependencies between singletons.  Say you have three singletons where singleton A depends on singleton B and uses it for something in it's initialization step.  Singleton B, similarly depends on singleton C.  If initialization is done in Awake, since unity doesn't guarantee an order in which Awake is called, this could lead to singleton A being initialized first, and throwing errors, similarly with singleton B and C.  If you instead follow the practice of initializing in this function, initialization is done on demand, allowing Awake to be called in any order and initialization to be auto sorted, as long as there are no cycles in your dependency graph.  It also allows for lazy initialization of singletons.

File: Assets/HoloToolkit/Utilities/Scripts/Performance/ViewportScaleManager.cs (hunk 6 above)
Comment (keluecke) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/ViewportScaleManager.cs]: Once again, I feel it is more useful to have fewer dependencies.  I would be fine with removing the voice commands, as they are simply an added utility for HoloLens.

File: Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs]: If that's the case, then should we have the Examples in the main repository?

File: Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs (hunk 3 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs]: Again this should be it's own Pull Request.

File: Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs (hunk 2 above)
Comment (keluecke) [reply Assets/HoloToolkit-Examples/Throwing/Scripts/ThrowingExampleSceneManager.cs]: I don't think so.  This scene is just a touch of code showing how one should consume controller information to take advantage of throwing.  Its sample code, not a component to use elsewhere.

File: Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs (hunk 3 above)
Comment (keluecke) [reply Assets/HoloToolkit-Examples/Utilities/Scripts/PlaneTargetGroupPicker.cs]: Ok, I can see that, I'll try to pull it out to its own pull request, and we can resolve that first then finish this one.

File: Assets/HoloToolkit/Utilities/Scripts/Throwing.cs (hunk 10 above)
Comment (keluecke) [reply Assets/HoloToolkit/Utilities/Scripts/Throwing.cs]: I'm confused by this.  It seems in order to 'correctly' wrap it, I would need to add a ton of #if-ed code, that is only destined to go out of date, this won't be in a version or MRTK that is intended for <2017.1.

File: Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs [hunk 11]
```diff
@@ -203,6 +205,10 @@ public void CalculateBoundaryVolume()
             // Ensuring that we set height of the bounds volume to be say 10 feet tall.
             boundaryBounds.Encapsulate(new Vector3(0, boundaryHeight, 0));
         }
+
+        protected override void InitializeInternal()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: This probably shouldn't be inside the block for `UNITY_2017_2_OR_NEWER`

Comment (keluecke) [reply Assets/HoloToolkit/Boundary/Scripts/BoundaryManager.cs]: Thats moved in this PR: https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/1518

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        private float lastSampleStartTime = 0;
+        private float lastFrameStartTime = 0;
+        private float gpuTimeAccumulator = 0;
+        private int currentFrame = 0;
+        private int currentSample = 0;
+
+        private TimeSampleStorage sampleStorage = null;
+        private LinkedList<TimeSample> timeSamples = new LinkedList<TimeSample>();
+
+        protected override void InitializeInternal()
```
Comment (timGerken) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: @StephenHodgson look here :)

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: Thanks Tim!

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+                TargetFrameRate = 60;
+                if (!Application.isEditor)
+                {
+                    Debug.LogWarning("Could not retrieve the HMD's native refresh rate. Assuming " + TargetFrameRate + " Hz.");
+                }
+            }
+
+            if (null == hmdCamera)
+            {
+                hmdCamera = Camera.main;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: nit: use `CameraCache.Main`

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+        protected override void InitializeInternal()
+        {
+            TargetFrameRate = (int)UnityEngine.XR.XRDevice.refreshRate;
+
+            if (TargetFrameRate == 0)
+            {
+                TargetFrameRate = 60;
+                if (!Application.isEditor)
+                {
+                    Debug.LogWarning("Could not retrieve the HMD's native refresh rate. Assuming " + TargetFrameRate + " Hz.");
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: Should we always log this if we're not in the Editor?

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceDisplay.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        UnityEngine.UI.Text fpsGUIText;
+        private bool show = true;
+        int frames = 0;
+        private float lastDisplayUpdate = 0;
+        private bool resetNextFrame = false;
+
+
+        void Awake()
+        {
+            fpsGUIText = GetComponent<UnityEngine.UI.Text>();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceDisplay.cs]: We should probably require this component for the class if we're always expecting to have it.

What about if someone uses `TextMesh` Instead?

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs (hunk 14 above)
Comment (keluecke) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: why not!

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: Just seems like it would always be the case, not the exception.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceCounters.cs]: Also, logs do incur some overhead, and this one is a warning, making me think there's something wrong.

File: Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceDisplay.cs (hunk 15 above)
Comment (keluecke) [reply Assets/HoloToolkit/Utilities/Scripts/Performance/PerformanceDisplay.cs]: I suppose they would need to modify the script slightly.  This is intended to be used mostly in the prefab is comes in.  I have added the require component, however.

**Comment by DDReaper** [comment]
Note, this PR is now stale due to the Namespace updated in the Dev_Working branch.  Please refresh or create a new PR against Dev_Working with these changes please.

**Comment by StephenHodgson** [comment]
@keluecke any update on this?

**Comment by StephenHodgson** [comment]
Closing this stale PR
